File: 402-移掉K位数字.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if len(num) == k:
            return '0'

        s = ''
        i = 0
        times = 0
        n = len(num)

        while times < n - k and n - i > n - k - times:
            current = float('inf')
            for j in range(i, k + times + 1):
                if int(num[j]) < current:         ## < 是为了处理相等元素，如果有相等的，取前者显然更好
                    i = j
                    current = int(num[j])
            s += num[i]
            i += 1
            times += 1
        
        if times < n - k:
            s += num[i:]

        i = 0
        while i < len(s) and s[i] == '0':
            i += 1

        return s[i:] if s[i:] else '0'

File: 402-移掉K位数字.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        if(num.size() == k)
            return "0";
        
        string s = "";
        int i = 0;
        int times = 0;
        int n = num.size();
        int current;

        while(times < n - k && n - i > n - k - times){
            current = INT_MAX;
            for(int j = i; j < k + times + 1; j++){
                if(static_cast<int>(num[j]) < current){
                    i = j;
                    current = static_cast<int>(num[j]);
                }
            }
            s += num[i];
            i += 1;
            times += 1;
        }

        if(times < n - k)
            s += num.substr(i);

        i = 0;
        while(i < s.size() && s[i] == '0')
            i++;
        s = s.substr(i);

        return s.empty() ? "0" : s;
    }
};
